azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -100,21 +97,23 @@ def complete(self, error=None):
         This process can be halted if a callback for the operation invokes the .halt_completion()
         method on this Operation.
 
+        Note that if an error is raised the operation may not be able to be completed.
+
         :param error: Optionally provide an Exception object indicating the error that caused
-            the completion. Providing an error indicates that the operation was unsucessful.
+            the completion. Providing an error indicates that the operation was unsuccessful.
+
+        :raises: OperationError if the operation cannot properly reach a completed state
+        :raises: OperationError if an error occurs in resolving any callbacks
         """
         if error:
             logger.debug("{}: completing with error {}".format(self.name, error))
         else:
             logger.debug("{}: completing without error".format(self.name))
 
         if self.completed or self.completing:
-            e = pipeline_exceptions.OperationError(
+            raise pipeline_exceptions.OperationError(
```

Comment:
>raise [](http://example.com/codeflow?start=12&length=5)

All of these raises introduce major flow changes. 
1. who catches them?
2. unlike the old code, which logged the exception and continued, the raise cause execution to stop and the stack to unwind.  This means that any code that might have previously been called will no longer be called. 

For example, look at ConnectionLockStage._unblock.  If one of the ops in old_queue hits this code, the rest of the ops in the queue will be lost.  

Raising here is downright destructive.